### ISSUE TITLE
listing available gpus

### ISSUE THREAD

**Issue by dougbevan** [OleHolmNielsen/Slurm_tools#3]
A very useful software.  How can we list the available vs used GRES for gpus?

For instance, if I do:

pestat -G

This is partially good, as I can see the GRES being used.  But it doesn't show the GRES available.

For CPUs, you get to see used/total (in my case 0/48).  How can I get a similar output for gpus?

**Comment by OleHolmNielsen** [comment]
I believe the sinfo command will give you the desired information about GRES in nodes.
For example, use this command:

sinfo -o "%P %G %D %N"

**Comment by OleHolmNielsen** [comment]
I understand now, so I've added a new column GRES/node which is printed if you select the -G flag. 
Can you try out the new script and tell me if this does what you want?

**Comment by dougbevan** [comment]
This is excellent.  I tried it on one of our single node systems, and I see the available gpu and the GRES/job.  Thanks for the addition -- this will be quite useful.

**Comment by cheekykite** [comment]
Hello.

Thanks for providing a good tool.

"GRES/job" is not showing up in a clustered environment.
Can I get an opinion?

```bash
master:pestat]#
master:pestat]# ./pestat  -G
GRES (Generic Resource) is printed after each jobid
Hostname       Partition     Node Num_CPU  CPUload  Memsize  Freemem  GRES/   Joblist
                            State Use/Tot              (MB)     (MB)  node    JobId User GRES/job ...
      n1        titanxp*     idle   0   6    0.07     60000    62869  gpu:TitanXP:2
      n2        titanxp*     idle   0   6    0.01     60000    62952  gpu:TitanXP:2
      n3        titanxp*     idle   0   6    0.01     60000    62860  gpu:TitanXP:2
      n4        titanxp*     idle   0   6    0.01     60000    62891  gpu:TitanXP:2
      n5        titanxp*     idle   0   6    0.01     60000    62971  gpu:TitanXP:2
      n6        titanxp*     idle   0   6    0.09     60000    62945  gpu:TitanXP:2
      n7        titanxp*     idle   0   6    0.01     60000    63096  gpu:TitanXP:2
      n8        titanxp*     idle   0   6    0.02     60000    63084  gpu:TitanXP:2
      n9        titanxp*      mix   4   6    2.39*    60000    49649  gpu:TitanXP:2 2367 sonic  2360 sonic
     n10        titanxp*     idle   0   6    0.01     60000    63082  gpu:TitanXP:2
master:pestat]#
master:pestat]#
master:pestat]# sinfo --version
slurm 18.08.8
master:pestat]#
master:pestat]# cat /etc/redhat-release
CentOS Linux release 7.8.2003 (Core)
master:pestat]#
master:pestat]#
```

**Comment by OleHolmNielsen** [comment]
You're running an old and obsolete version of Slurm. Later versions have significantly improved GPU support, so maybe that's why you don't get the expected information.

The pestat command obtains information from Slurm with:
   sinfo -h -N $partition $hostlist $statelist -o "%N %P %C %O %m %e %t %Z %G" 
where the %G option prints:
     %G    Generic resources (gres) associated with the nodes.
Please check "man sinfo" in your Slurm version to see if %G exists.

**Comment by OleHolmNielsen** [comment]
Can you please test the latest version of pestat?  The GRES/job is now being printed correctly.

**Comment by clue2** [comment]
It might be helpful to change the formatting from -o to -O to make use of the extra formatting options (such as GresUsed)

sinfo -h -N $partition $hostlist $statelist -o "%N %P %C %O %m %e %t %Z %G"

becomes:

sinfo -h -N $partition $hostlist $statelist -O "Nodes,Partition,CPUsState,CPUsLoad,Memory,FreeMem,StateCompact,Threads,Gres"
![Screen Shot 2022-08-17 at 2 18 05 pm](https://user-images.githubusercontent.com/30360870/185033848-65a7ac4c-edec-4208-b87e-e8a6e3cfcdcf.png)

You could then add in GresUsed ( ideally cleaning it up a bit ) to achieve a more helpful overview of how many GPUs are in use/available on a node

![Screen Shot 2022-08-17 at 2 18 16 pm](https://user-images.githubusercontent.com/30360870/185033880-9a5423ba-11c8-47db-a564-27930aa14e11.png)

**Comment by yzs981130** [comment]
> change the formatting from -o to -O

https://github.com/OleHolmNielsen/Slurm_tools/blob/21ef8a6852b08b94ff004e9ab6ad6065376fce21/pestat/pestat#L340

I believe `pestat` has already used `-O` to retrieve information.

I am experiencing the same problem with you, to add a node-level `GresUsed` in the output of `pestat`. Therefore, I added it in my personal fork: https://github.com/yzs981130/Slurm_tools/commit/7e711af416d28bd98c1e9af71a23d686b3a10044. Hope it can help you!

cc @OleHolmNielsen  What do you think about the node-level `GresUsed`? Since it is my first time using `awk`, I could send a draft pr if you think it is also needed.

**Comment by clue2** [comment]
You're right, it does use `-O` now - I hadn't actually checked the code & was just going by the comments above. Thanks!

In that case, just changing `Gres` to `GresUsed` does a good enough job
![Screen Shot 2022-08-22 at 10 51 13 am](https://user-images.githubusercontent.com/30360870/185819497-6f75d205-d877-4f8a-bbf4-3de5611e51a7.png)

In your fork the formatting has become a bit off for me:
![Screen Shot 2022-08-22 at 10 49 41 am](https://user-images.githubusercontent.com/30360870/185819268-c10a974c-3777-4f29-8efd-1d586ed45502.png)

**Comment by OleHolmNielsen** [comment]
Thank you for your suggestion.
The GRES output shows how many GPUs are physically in the node.  

With "pestat -G" the GRES used by each job on the node is printed. One could count manually how many GPUs are used.

I agree that the "sinfo -O GRESUSED" gives a useful summary of how many GPUs are in use.

However, I think that printing both GRES and GRESUSED data makes the output very long and difficult to read.

Maybe one could think of simplifying by having a "Num_GPU" column with simply the "Use/Tot" numbers.  Some complicated parsing of GRES and GRESUSED would be needed. 

There could be non-GPU types of GRES, see https://slurm.schedmd.com/gres.conf.html

Do you have suggestions for making the output of pestat more useful and simple to read?

**Comment by OleHolmNielsen** [comment]
Note added: Sites have to define their own GRES types in slurm.conf using the GresTypes parameter.
It can become complex for "pestat" to decode all possible GRES types and extract numbers for "Use/Tot".
